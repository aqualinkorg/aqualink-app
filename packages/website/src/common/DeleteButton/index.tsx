import React, { useState } from 'react';
import { IconButton, Collapse, LinearProgress } from '@mui/material';
import { WithStyles } from '@mui/styles';
import withStyles from '@mui/styles/withStyles';
import createStyles from '@mui/styles/createStyles';
import Alert from '@mui/material/Alert';
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline';
import CloseIcon from '@mui/icons-material/Close';
import DeleteDialog, { Action } from '../Dialog';

const DeleteButton = ({
  header,
  content = null,
  onConfirm,
  onSuccess = () => {},
  onError = () => {},
  classes,
}: DeleteButtonProps) => {
  const [open, setOpen] = useState<boolean>(false);
  const [alertOpen, setAlertOpen] = useState<boolean>(false);
  const [alertText, setAlertText] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
    setAlertOpen(false);
    setAlertText('');
  };

  const onDelete = async () => {
    setLoading(true);
    try {
      await onConfirm();
      onSuccess?.();
    } catch (error) {
      onError?.();
      setAlertOpen(true);
      setAlertText((error as any)?.message);
    }
    setLoading(false);
  };

  const dialogActions: Action[] = [
    {
      size: 'small',
      variant: 'outlined',
      color: 'secondary',
      text: 'Cancel',
      action: handleClose,
    },
    {
      size: 'small',
      variant: 'outlined',
      color: 'primary',
      text: 'Yes',
      action: onDelete,
    },
  ];

  return (
    <>
      <IconButton onClick={handleClickOpen} size="large">
        <DeleteOutlineIcon color="primary" />
      </IconButton>
      <DeleteDialog
        open={open}
        onClose={handleClose}
        header={header}
        content={
          <>
            {content}
            {loading && <LinearProgress />}
            <Collapse className={classes.alert} in={alertOpen}>
              <Alert
                severity="error"
                action={
                  <IconButton
                    color="inherit"
                    size="small"
                    onClick={() => {
                      setAlertOpen(false);
                    }}
                  >
                    <CloseIcon fontSize="inherit" />
                  </IconButton>
                }
              >
                {alertText}
              </Alert>
            </Collapse>
          </>
        }
        actions={dialogActions}
      />
    </>
  );
};

const styles = () =>
  createStyles({
    alert: {
      width: '100%',
    },
  });

interface DeleteButtonIncomingProps {
  header?: string;
  content?: JSX.Element | null;
  onConfirm: () => Promise<any>;
  onSuccess?: () => void;
  onError?: () => void;
}

type DeleteButtonProps = DeleteButtonIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(DeleteButton);
