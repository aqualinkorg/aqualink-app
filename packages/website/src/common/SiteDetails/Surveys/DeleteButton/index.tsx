import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  withStyles,
  WithStyles,
  createStyles,
  IconButton,
  Collapse,
  LinearProgress,
} from "@material-ui/core";
import Alert from "@material-ui/lab/Alert";
import DeleteOutlineIcon from "@material-ui/icons/DeleteOutline";
import CloseIcon from "@material-ui/icons/Close";
import moment from "moment";

import surveyServices from "../../../../services/surveyServices";
import { userInfoSelector } from "../../../../store/User/userSlice";
import { surveysRequest } from "../../../../store/Survey/surveyListSlice";
import DeleteDialog, { Action } from "../../../Dialog";

const DeleteButton = ({
  reefId,
  surveyId,
  diveDate,
  classes,
}: DeleteButtonProps) => {
  const user = useSelector(userInfoSelector);
  const dispatch = useDispatch();
  const [open, setOpen] = useState<boolean>(false);
  const [alertOpen, setAlertOpen] = useState<boolean>(false);
  const [alertText, setAlertText] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
    setAlertOpen(false);
    setAlertText("");
  };

  const onSurveyDelete = () => {
    if (surveyId && user && user.token) {
      setLoading(true);
      surveyServices
        .deleteSurvey(reefId, surveyId, user.token)
        .then(() => {
          dispatch(surveysRequest(`${reefId}`));
        })
        .catch((error) => {
          setAlertOpen(true);
          setAlertText(error.message);
        })
        .finally(() => setLoading(false));
    }
  };

  const dialogActions: Action[] = [
    {
      size: "small",
      variant: "contained",
      color: "secondary",
      text: "No",
      action: handleClose,
    },
    {
      size: "small",
      variant: "contained",
      color: "primary",
      text: "Yes",
      action: onSurveyDelete,
    },
  ];

  return (
    <>
      <IconButton onClick={handleClickOpen}>
        <DeleteOutlineIcon color="primary" />
      </IconButton>
      <DeleteDialog
        open={open}
        onClose={handleClose}
        header={`Are you sure you would like to delete the survey for ${moment(
          diveDate
        ).format(
          "MM/DD/YYYY"
        )}? It will delete all media assosciated with this survey.`}
        content={
          <>
            {loading && <LinearProgress />}
            <Collapse className={classes.alert} in={alertOpen}>
              <Alert
                severity="error"
                action={
                  <IconButton
                    color="inherit"
                    size="small"
                    onClick={() => {
                      setAlertOpen(false);
                    }}
                  >
                    <CloseIcon fontSize="inherit" />
                  </IconButton>
                }
              >
                {alertText}
              </Alert>
            </Collapse>
          </>
        }
        actions={dialogActions}
      />
    </>
  );
};

const styles = () =>
  createStyles({
    alert: {
      width: "100%",
    },
  });

interface DeleteButtonIncomingProps {
  surveyId?: number | null;
  diveDate?: string | null;
  reefId: number;
}

DeleteButton.defaultProps = {
  surveyId: null,
  diveDate: null,
};

type DeleteButtonProps = DeleteButtonIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(DeleteButton);
