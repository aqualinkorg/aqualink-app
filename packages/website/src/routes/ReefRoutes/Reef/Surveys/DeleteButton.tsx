import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  withStyles,
  WithStyles,
  createStyles,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  IconButton,
  Button,
  Collapse,
  LinearProgress,
} from "@material-ui/core";
import Alert from "@material-ui/lab/Alert";
import DeleteOutlineIcon from "@material-ui/icons/DeleteOutline";
import CloseIcon from "@material-ui/icons/Close";
import moment from "moment";

import surveyServices from "../../../../services/surveyServices";
import { userInfoSelector } from "../../../../store/User/userSlice";
import { surveysRequest } from "../../../../store/Survey/surveyListSlice";

const DeleteButton = ({
  reefId,
  surveyId,
  diveDate,
  classes,
}: DeleteButtonProps) => {
  const user = useSelector(userInfoSelector);
  const dispatch = useDispatch();
  const [open, setOpen] = useState<boolean>(false);
  const [alertOpen, setAlertOpen] = useState<boolean>(false);
  const [alertText, setAlertText] = useState<string>("");
  const [loading, setLoading] = useState<boolean>(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
    setAlertOpen(false);
    setAlertText("");
  };

  const onSurveyDelete = () => {
    if (surveyId && user && user.token) {
      setLoading(true);
      surveyServices
        .deleteSurvey(surveyId, user.token)
        .then(() => {
          dispatch(surveysRequest(`${reefId}`));
        })
        .catch((error) => {
          setAlertOpen(true);
          setAlertText(error.message);
        })
        .finally(() => setLoading(false));
    }
  };

  return (
    <>
      <IconButton onClick={handleClickOpen}>
        <DeleteOutlineIcon color="primary" />
      </IconButton>
      <Dialog open={open} onClose={handleClose}>
        <DialogTitle className={classes.dialogTitle}>
          <Typography color="textSecondary">
            Are you sure you would like to delete the survey for{" "}
            {moment(diveDate).format("MM/DD/YYYY")}?
          </Typography>
        </DialogTitle>
        <DialogContent className={classes.dialogContent}>
          {loading && <LinearProgress />}
          <Collapse className={classes.alert} in={alertOpen}>
            <Alert
              severity="error"
              action={
                <IconButton
                  color="inherit"
                  size="small"
                  onClick={() => {
                    setAlertOpen(false);
                  }}
                >
                  <CloseIcon fontSize="inherit" />
                </IconButton>
              }
            >
              {alertText}
            </Alert>
          </Collapse>
        </DialogContent>
        <DialogActions>
          <Button
            size="small"
            variant="contained"
            color="secondary"
            onClick={handleClose}
          >
            No
          </Button>
          <Button
            size="small"
            variant="contained"
            color="primary"
            onClick={onSurveyDelete}
          >
            Yes
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

const styles = () =>
  createStyles({
    dialogTitle: {
      backgroundColor: "#F4F4F4",
    },
    dialogContent: {
      padding: 0,
    },
    alert: {
      width: "100%",
    },
  });

interface DeleteButtonIncomingProps {
  surveyId?: number;
  diveDate?: string | null;
  reefId: number;
}

type DeleteButtonProps = DeleteButtonIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(DeleteButton);
