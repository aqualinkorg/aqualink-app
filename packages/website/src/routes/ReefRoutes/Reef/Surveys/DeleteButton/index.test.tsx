import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";
import DeleteButton from ".";

const mockStore = configureStore([]);

describe("Surveys Delete Button", () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      user: {
        userInfo: {
          email: "test@mail.com",
          fullName: "Test User",
          adminLevel: "super_admin",
          firebaseUid: "RaNdOmStRiNg",
          administeredReefs: [],
          token: "RaNdOmStRiNg",
        },
      },
    });

    store.dispatch = jest.fn();

    element = render(
      <Provider store={store}>
        <DeleteButton reefId={0} surveyId={0} diveDate="10/2/2020" />
      </Provider>
    ).container;
  });

  it("should render with given state from Redux store", () => {
    expect(element).toMatchSnapshot();
  });
});
