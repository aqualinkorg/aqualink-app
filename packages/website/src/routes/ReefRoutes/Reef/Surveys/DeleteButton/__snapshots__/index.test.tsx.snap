// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surveys Delete Button should render with given state from Redux store 1`] = `
<div>
  <mock-iconbutton>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
      />
    </svg>
  </mock-iconbutton>
  <mock-dialog
    open="false"
  >
    <mock-dialogtitle
      classname="Dialog-dialogTitle-2"
    >
      <mock-typography>
        Are you sure you would like to delete the survey for 10/02/2020? It will delete all media assosciated with this survey.
      </mock-typography>
    </mock-dialogtitle>
    <mock-dialogcontent>
      <div
        class="MuiCollapse-container DeleteButton-alert-1 MuiCollapse-hidden"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner"
          >
            <div
              class="MuiPaper-root MuiAlert-root MuiAlert-standardError MuiPaper-elevation0"
              role="alert"
            >
              <div
                class="MuiAlert-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
              </div>
              <div
                class="MuiAlert-message"
              />
              <div
                class="MuiAlert-action"
              >
                <mock-iconbutton
                  color="inherit"
                  size="small"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </mock-iconbutton>
              </div>
            </div>
          </div>
        </div>
      </div>
    </mock-dialogcontent>
    <mock-dialogactions>
      <mock-button
        color="secondary"
        size="small"
        variant="contained"
      >
        No
      </mock-button>
      <mock-button
        color="primary"
        size="small"
        variant="contained"
      >
        Yes
      </mock-button>
    </mock-dialogactions>
  </mock-dialog>
</div>
`;
